### PR TITLE
verify: config is encoded in utf8; cwd is absolute path; env is valid

### LINKED ISSUE #67: Where do we enforce UTF-8 JSON?
Spun off from #66, where “All configuration JSON MUST be encoded in UTF-8” is checked.  And it seems to be:

    $ ./ocitools generate
    $ iconv -f UTF-8 -t UTF-16 config.json -o config-16.json
    $ mv config-16.json config.json
    $ file config.json
    config.json: Little-endian UTF-16 Unicode text
    $ ./ocitools validate --path .
    FATA[0000] invalid character 'ÿ' looking for beginning of value

But I can't find [Go docs on what input encodings are supported][1] or UTF references in ocitools:

    $ git grep -i utf origin/master
    Binary file origin/master:rootfs.tar.gz matches

The Go docs [reference RFC 4627][1], and the RFC [explains how to detect the encoding][2], but maybe Go doesn't support that yet?  Can someone with more Go+JSON experience explain how this works?

[1]: https://golang.org/pkg/encoding/json/
[2]: https://tools.ietf.org/html/rfc4627#section-3

### PR DESCRIPTION
**Pull request by liangchenye**
fix https://github.com/opencontainers/ocitools/issues/67
Signed-off-by: liang chenye <liangchenye@huawei.com>

### CODE REVIEW

**Comment by liangchenye** [comment]
@mrunalp @wking PTAL

File: validate.go [hunk 1]
```diff
@@ -129,16 +132,26 @@ func checkPlatform(platform rspec.Platform) {
 }
 
 func checkProcess(process rspec.Process, rootfs string) {
+	if !path.IsAbs(process.Cwd) {
+		logrus.Fatalf("Cwd '%s' is not an absolute path", process.Cwd)
```
Comment (mrunalp) [inline validate.go]: you can use %q instead of '%s'

File: validate.go [hunk 2]
```diff
@@ -129,16 +132,26 @@ func checkPlatform(platform rspec.Platform) {
...
+	}
+
+	for _, env := range process.Env {
+		if !envValid(env) {
+			logrus.Fatalf("Env '%s' should be in the form of 'key=value'. The left hand side must consist solely of letters, digits, and underscores '_'.", env)
```
Comment (mrunalp) [inline validate.go]: same here as well and elsewhere in the PR.

**Comment by liangchenye** [comment]
Thanks @mrunalp , updated.

File: validate.go [hunk 3]
```diff
@@ -53,23 +56,23 @@ var bundleValidateCommand = cli.Command{
...
-		sf, err := os.Open(path.Join(inputPath, "config.json"))
+		content, err := ioutil.ReadFile(path.Join(inputPath, "config.json"))
 		if err != nil {
 			logrus.Fatal(err)
 		}
-
-		defer sf.Close()
-
+		if !utf8.Valid(content) {
+			logrus.Fatalf("'Config.json' is not encoded in UTF-8")
```
Comment (wking) [inline validate.go]: `Config.json` → `config.json` for folks on case-sensitive filesystems?

Comment (wking) [inline validate.go]: Actually, maybe store the `path.Join` in a variable, and pass that variable to both `ReadFile` and this error message.

File: validate.go [hunk 4]
```diff
@@ -214,11 +225,24 @@ func checkSeccomp(s rspec.Seccomp) {
...
+			logrus.Fatalf("Seccomp.Architecture %q is invalid", s.Architectures[index])
 		}
 	}
 }
 
+func envValid(env string) bool {
+	items := strings.Split(env, "=")
+	if len(items) != 2 {
+		return false
+	}
```
Comment (wking) [inline validate.go]: Multiple equals are valid (e.g. `LS_COLORS=rs=0:di=01;34:…`).

File: validate.go [hunk 5]
```diff
@@ -122,23 +125,33 @@ func checkPlatform(platform rspec.Platform) {
...
+			logrus.Fatalf("Combination of %q and %q is invalid.", platform.OS, platform.Arch)
 		}
 	}
-	logrus.Fatalf("Operation system '%s' of the bundle is not supported yet.", platform.OS)
+	logrus.Fatalf("Operation system %q of the bundle is not supported yet.", platform.OS)
 }
 
 func checkProcess(process rspec.Process, rootfs string) {
+	if !path.IsAbs(process.Cwd) {
+		logrus.Fatalf("Cwd %q is not an absolute path", process.Cwd)
```
Comment (mrunalp) [inline validate.go]: Cwd -> cwd to match what is expected to be the property name in config.json

**Comment by liangchenye** [comment]
@wking @mrunalp  Thanks, PR updated according to your comments.

File: validate.go [hunk 6]
```diff
@@ -89,17 +93,17 @@ func bundleValidate(spec rspec.Spec, rootfs string) {
...
 
 func checkMounts(mounts []rspec.Mount, rootfs string) {
 	for _, mount := range mounts {
 		rootfsPath := path.Join(rootfs, mount.Destination)
 		if fi, err := os.Stat(rootfsPath); err != nil {
-			logrus.Fatalf("Cannot find the mount point: %v", rootfsPath)
+			logrus.Fatalf("Cannot find the mount destination %q", rootfsPath)
 		} else if !fi.IsDir() {
-			logrus.Fatalf("Mount point: %v is not a directory.", rootfsPath)
+			logrus.Fatalf("Mount destination %q is not a directory.", rootfsPath)
```
Comment (wking) [inline validate.go]: It predates this PR, but `rootfsPath` is a strange name for a variable holding the mount destination.

**Comment by wking** [comment]
On Thu, May 19, 2016 at 01:24:03AM -0700, 梁辰晔 (Liang Chenye) wrote:
> @wking @mrunalp  Thanks, PR updated according to your comments.

664508c looks good to me.  I think we should rename a variable [1],
but I'm fine with that happening in a follow-up PR.

[1]: https://github.com/opencontainers/ocitools/pull/70#discussion_r63911041

**Comment by liangchenye** [comment]
@wking  corrected

**Comment by wking** [comment]
On Thu, May 19, 2016 at 07:50:48PM -0700, 梁辰晔 (Liang Chenye) wrote:
>  @wking  corrected

01c8d1d looks lovely :).

**Comment by mrunalp** [comment]
LGTM

**Pull request merged**